flow/layers/clip_shape_layer.h

```diff
@@ -0,0 +1,112 @@
+    DiffContext::AutoSubtreeRestore subtree(context);
+    auto* prev = static_cast<const ClipShapeLayer<ClipShape>*>(old_layer);
+    if (!context->IsSubtreeDirty()) {
+      FML_DCHECK(prev);
+      if (clip_behavior_ != prev->clip_behavior_ ||
+          clip_shape_ != prev->clip_shape_) {
+        context->MarkSubtreeDirty(context->GetOldLayerPaintRegion(old_layer));
+      }
+    }
+    if (context->PushCullRect(clip_shape_bounds())) {
+      DiffChildren(context, prev);
+    }
+    context->SetLayerPaintRegion(this, context->CurrentSubtreeRegion());
+  }
+
+  void Preroll(PrerollContext* context, const SkMatrix& matrix) override {
+    SkRect previous_cull_rect = context->cull_rect;
+    if (!context->cull_rect.intersect(clip_shape_bounds())) {
+      context->cull_rect.setEmpty();
+    }
+    Layer::AutoPrerollSaveLayerState save =
+        Layer::AutoPrerollSaveLayerState::Create(context, UsesSaveLayer());
+    OnMutatorsStackPushClipShape(context->mutators_stack);
+
+    // Collect inheritance information on our children in Preroll so that
+    // we can pass it along by default.
+    context->subtree_can_inherit_opacity = true;
+
+    SkRect child_paint_bounds = SkRect::MakeEmpty();
+    PrerollChildren(context, matrix, &child_paint_bounds);
+    if (child_paint_bounds.intersect(clip_shape_bounds())) {
+      set_paint_bounds(child_paint_bounds);
+    }
+
+    // If we use a SaveLayer then we can accept opacity on behalf
+    // of our children and apply it in the saveLayer.
+    if (UsesSaveLayer()) {
+      context->subtree_can_inherit_opacity = true;
+    }
+
+    context->mutators_stack.Pop();
+    context->cull_rect = previous_cull_rect;
+  }
+
+  void Paint(PaintContext& context) const override {
+    FML_DCHECK(needs_painting(context));
+
+    SkAutoCanvasRestore save(context.internal_nodes_canvas, true);
+    OnCanvasClipShape(context.internal_nodes_canvas);
+
+    if (!UsesSaveLayer()) {
+      PaintChildren(context);
+      return;
+    }
+
+    AutoCachePaint cache_paint(context);
+    TRACE_EVENT0("flutter", "Canvas::saveLayer");
+    context.internal_nodes_canvas->saveLayer(paint_bounds(),
+                                             cache_paint.paint());
+
+    PaintChildren(context);
+
+    context.internal_nodes_canvas->restore();
+    if (context.checkerboard_offscreen_layers) {
+      DrawCheckerboard(context.internal_nodes_canvas, paint_bounds());
+    }
+  }
+
+  bool UsesSaveLayer() const {
+    return clip_behavior_ == Clip::antiAliasWithSaveLayer;
+  }
+
+ protected:
+  virtual const SkRect& clip_shape_bounds() const = 0;
+  virtual void OnMutatorsStackPushClipShape(MutatorsStack& mutators_stack) = 0;
+  virtual void OnCanvasClipShape(SkCanvas* canvas) const = 0;
+  virtual ~ClipShapeLayer() = default;
+
+  const ClipShape& clip_shape() const { return clip_shape_; }
+  Clip clip_behavior() const { return clip_behavior_; }
+
+ private:
+  const ClipShape clip_shape_;
+  Clip clip_behavior_;
```

Comment:
Should these just be protected? They came from the subclass anyway and it would eliminate the need for the getters.

Reply:
Since the google c++ style guide recommends making all data members private, maybe we should keep it? At the same time, I guess the getter should be treated as an inline function by the compiler. WDYT?

c.f.
https://google.github.io/styleguide/cppguide.html#Access_Control